File: algorithm/src/js/03_quick_sort.js
/**
 * 快速排序 时间复杂度为O(n*logn)
 * @param {*} array 
 */
const quick_sort = array => {
  if (array.length < 2) {
    return array;
  } else {
    // 随意选取一个值为基准值
    let pivot = array[0];
    let less = array.slice(1).filter(item => item <= pivot);
    let greater = array.slice(1).filter(item => item > pivot);
    return [...quick_sort(less), pivot, ...quick_sort(greater)];
  }
};

console.log(quick_sort([3, 7, 9, 12, 68, 34, 21]));

File: algorithm/src/python/03_quick_sort.py
# coding=utf-8


# 快速排序
def quick_sort(list):
    if len(list) < 2:
        return list
    else:
        pivot = list[0]
        less = [i for i in list[1:] if i <= pivot]
        greater = [i for i in list[1:] if i > pivot]
        return quick_sort(less) + [pivot] + quick_sort(greater)


print(quick_sort([5, 3, 6, 2, 10]))
